Differences in struct layout between memory and storage

ritzdorf opened the pull request:
More explicit description with an example.

mijovic commented on docs/internals/layout_in_memory.rst:
```diff
@@ -20,6 +20,17 @@ memory is never freed (this might change in the future).
 
 Elements in memory arrays in Solidity always occupy multiples of 32 bytes (this
 is even true for ``byte[]``, but not for ``bytes`` and ``string``).
+This implies that structs can take up more space in memory than they do in 
```
```suggestion
This implies that structs can take up more space in memory than they do in
```

mijovic commented:
Also here documentation is mentioning this is true for memory arrays, so I think you would need to update your example a little bit.

ritzdorf commented:
> Also here documentation is mentioning this is true for memory arrays, so I think you would need to update your example a little bit.

I wanted to make an example for structs, as this is a question I have often heard about. Happy to add one for arrays too. Is that what you mean @mijovic ?

mijovic commented:
> > Also here documentation is mentioning this is true for memory arrays, so I think you would need to update your example a little bit.
> 
> I wanted to make an example for structs, as this is a question I have often heard about. Happy to add one for arrays too. Is that what you mean @mijovic ?

I am not sure if this is best place for example about structs. But, in general this example can be a bit misleading. Storage is divided to slots of 256 bytes. In case of your struct, it will occupy one slot.
The only way to use reminder of that slot is via inline assembly.

But definitely, documentation lacks info on layouts of different types.

I suggest moving this to different location in documentation and using a slight different wording.

ritzdorf commented:
>  Storage is divided to slots of 256 bytes. 

I guess you meant 256 bit = 32 bytes.

>  In case of your struct, it will occupy one slot.

The struct occupies 3 slots.

> I suggest moving this to different location in documentation and using a slight different wording.

I moved it to a different subsection.

Thanks for your feedback @mijovic .

chriseth commented on docs/internals/layout_in_memory.rst:
```diff
@@ -36,4 +36,41 @@ elements.
+
+
+
+.. index: calldata layout
```
It seems this line belong to the top of `layout_in_calldata` - can you move it there, please?

ritzdorf replied:
Thanks @chriseth ! Fixed in 8a2731bfdb32aa81a84d6da3ef211958073bae4a

chriseth commented on docs/internals/layout_in_memory.rst:
```diff
@@ -36,4 +36,41 @@ elements.
+
+::
+
+    struct S{
```
```suggestion
    struct S {
```

mijovic commented:
@ritzdorf Sorry for completely off review. I wasn't fully concentrated.
Thanks for moving it into different section. After fixing few comments on this PR from Chris, I guess it would be good to merge.

mijovic commented:
Can you squash into one commit?

ritzdorf committed:
Differences in layout between memory and storage

- More explicit description with two examples
- Moved index pointer to correct file

ritzdorf commented:
> Can you squash into one commit?

Done. @mijovic
